rclcpp/test/rclcpp/test_generic_pubsub.cpp

```diff
@@ -0,0 +1,185 @@
+  {
+    auto publisher = publisher_node_->create_publisher<test_msgs::msg::Strings>(topic, 10);
+    publishers_.push_back(publisher);
+  }
+
+  std::vector<std::string> subscribe_raw_messages(
+    size_t expected_messages_number, const std::string & topic_name, const std::string & type)
+  {
+    std::vector<std::string> messages;
+    size_t counter = 0;
+    auto subscription = node_->create_generic_subscription(
+      topic_name, type, rclcpp::QoS(1),
+      [&counter, &messages](std::shared_ptr<rclcpp::SerializedMessage> message) {
+        test_msgs::msg::Strings string_message;
+        rclcpp::Serialization<test_msgs::msg::Strings> serializer;
+        serializer.deserialize_message(message.get(), &string_message);
+        messages.push_back(string_message.string_value);
+        counter++;
+      });
+
+    while (counter < expected_messages_number) {
+      rclcpp::spin_some(node_);
+    }
+    return messages;
+  }
+
+  rclcpp::SerializedMessage serialize_string_message(const std::string & message)
+  {
+    test_msgs::msg::Strings string_message;
+    string_message.string_value = message;
+    rclcpp::Serialization<test_msgs::msg::Strings> ser;
+    SerializedMessage result;
+    ser.serialize_message(&string_message, &result);
+    return result;
+  }
+
+  void sleep_to_allow_topics_discovery()
+  {
+    // This is a short sleep to allow the node some time to discover the topic
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+  }
+
+  template<typename Condition, typename Duration>
+  bool wait_for(const Condition & condition, const Duration & timeout)
+  {
+    using clock = std::chrono::system_clock;
+    auto start = clock::now();
+    while (!condition()) {
+      if ((clock::now() - start) > timeout) {
+        return false;
+      }
+      rclcpp::spin_some(node_);
+    }
+    return true;
+  }
+
+  std::shared_ptr<rclcpp::Node> node_;
+  rclcpp::Node::SharedPtr publisher_node_;
+  std::vector<std::shared_ptr<rclcpp::PublisherBase>> publishers_;
+};
+
+
+TEST_F(RclcppGenericNodeFixture, publisher_and_subscriber_work)
+{
+  // We currently publish more messages because they can get lost
+  std::vector<std::string> test_messages = {"Hello World", "Hello World"};
+  std::string topic_name = "string_topic";
+  std::string type = "test_msgs/Strings";
+
+  auto publisher = node_->create_generic_publisher(
+    topic_name, type, rclcpp::QoS(1));
+
+  auto subscriber_future_ = std::async(
+    std::launch::async, [this, topic_name, type] {
+      return subscribe_raw_messages(1, topic_name, type);
+    });
+  // Give time to the subscriber to start.
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  for (const auto & message : test_messages) {
+    publisher->publish(serialize_string_message(message));
+    // This is necessary because sometimes, the subscriber is initialized very late
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
```

Comment:
These sleeps are pretty brittle. It would be better to wait until we actually detect the subscriptions are ready (or matched with the publisher). I'm working on an API to help with this, as it is a common pattern. Maybe it's worth leaving a TODO here and we can come back and refactor later.